In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
%matplotlib inline

from decimal import Decimal, ROUND_HALF_UP

In [2]:
# Numpy
import numpy as np

# Dataframe operations
import pandas as pd

# Data visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

# Scalers
from sklearn.preprocessing import StandardScaler

# Data Split
from sklearn.model_selection import train_test_split

# Cross-validation
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import LeaveOneOut
from sklearn.model_selection import ShuffleSplit
from sklearn.model_selection import StratifiedShuffleSplit

# Linear-models
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVR
from sklearn.svm import LinearSVC

# # Dicision Tree
# import pydotplus
# from sklearn.tree import DecisionTreeRegressor
# from sklearn.tree import DecisionTreeClassifier
# from sklearn.tree import export_graphviz
# from sklearn import datasets
# from IPython.display import Image

# Random forest
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import RandomForestClassifier

# Evaluation
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_squared_error

# Tuning parameters
from sklearn.model_selection import GridSearchCV

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/ensemble/weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d


Kaggleの Home Credit Default Risk コンペティションを題材に、機械学習の実践的な流れを学びます。特に適切な 検証 を行い、高い 汎化性能 のあるモデルを完成させることを目指します。

In [3]:
df = pd.read_csv('application_train.csv')
display(df)

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0           100002       1         Cash loans           M            N   
1           100003       0         Cash loans           F            N   
2           100004       0    Revolving loans           M            Y   
3           100006       0         Cash loans           F            N   
4           100007       0         Cash loans           M            N   
5           100008       0         Cash loans           M            N   
6           100009       0         Cash loans           F            Y   
7           100010       0         Cash loans           M            Y   
8           100011       0         Cash loans           F            N   
9           100012       0    Revolving loans           M            N   
10          100014       0         Cash loans           F            N   
11          100015       0         Cash loans           F            N   
12          100016       0         Cash loans           F            N   
13          100017       0         Cash loans           M            Y   
14          100018       0         Cash loans           F            N   
15          100019       0         Cash loans           M            Y   
16          100020       0         Cash loans           M            N   
17          100021       0    Revolving loans           F            N   
18          100022       0    Revolving loans           F            N   
19          100023       0         Cash loans           F            N   
20          100024       0    Revolving loans           M            Y   
21          100025       0         Cash loans           F            Y   
22          100026       0         Cash loans           F            N   
23          100027       0         Cash loans           F            N   
24          100029       0         Cash loans           M            Y   
25          100030       0         Cash loans           F            N   
26          100031       1         Cash loans           F            N   
27          100032       0         Cash loans           M            N   
28          100033       0         Cash loans           M            Y   
29          100034       0    Revolving loans           M            N   
...            ...     ...                ...         ...          ...   
307481      456225       1         Cash loans           M            N   
307482      456226       0         Cash loans           F            Y   
307483      456227       0         Cash loans           F            N   
307484      456228       0         Cash loans           F            Y   
307485      456229       0    Revolving loans           F            N   
307486      456230       0         Cash loans           F            Y   
307487      456231       0         Cash loans           M            N   
307488      456232       0         Cash loans           F            N   
307489      456233       1         Cash loans           F            N   
307490      456234       0         Cash loans           M            N   
307491      456235       0         Cash loans           M            Y   
307492      456236       0         Cash loans           M            Y   
307493      456237       0         Cash loans           F            N   
307494      456238       0         Cash loans           M            Y   
307495      456239       0         Cash loans           M            Y   
307496      456240       0    Revolving loans           F            N   
307497      456241       0         Cash loans           F            Y   
307498      456242       0         Cash loans           M            Y   
307499      456243       0         Cash loans           F            N   
307500      456244       0         Cash loans           F            N   
307501      456245       0         Cash loans           F            N   
307502      456246       0         Cash loans           F            N   
307503      

In [4]:
X_original = df.iloc[:, :].drop(['TARGET'], axis=1)
y_original = df.loc[:, 'TARGET']
display(X_original)
display(y_original)

SK_ID_CURR NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0           100002         Cash loans           M            N   
1           100003         Cash loans           F            N   
2           100004    Revolving loans           M            Y   
3           100006         Cash loans           F            N   
4           100007         Cash loans           M            N   
5           100008         Cash loans           M            N   
6           100009         Cash loans           F            Y   
7           100010         Cash loans           M            Y   
8           100011         Cash loans           F            N   
9           100012    Revolving loans           M            N   
10          100014         Cash loans           F            N   
11          100015         Cash loans           F            N   
12          100016         Cash loans           F            N   
13          100017         Cash loans           M            Y   
14          100018         Cash loans           F            N   
15          100019         Cash loans           M            Y   
16          100020         Cash loans           M            N   
17          100021    Revolving loans           F            N   
18          100022    Revolving loans           F            N   
19          100023         Cash loans           F            N   
20          100024    Revolving loans           M            Y   
21          100025         Cash loans           F            Y   
22          100026         Cash loans           F            N   
23          100027         Cash loans           F            N   
24          100029         Cash loans           M            Y   
25          100030         Cash loans           F            N   
26          100031         Cash loans           F            N   
27          100032         Cash loans           M            N   
28          100033         Cash loans           M            Y   
29          100034    Revolving loans           M            N   
...            ...                ...         ...          ...   
307481      456225         Cash loans           M            N   
307482      456226         Cash loans           F            Y   
307483      456227         Cash loans           F            N   
307484      456228         Cash loans           F            Y   
307485      456229    Revolving loans           F            N   
307486      456230         Cash loans           F            Y   
307487      456231         Cash loans           M            N   
307488      456232         Cash loans           F            N   
307489      456233         Cash loans           F            N   
307490      456234         Cash loans           M            N   
307491      456235         Cash loans           M            Y   
307492      456236         Cash loans           M            Y   
307493      456237         Cash loans           F            N   
307494      456238         Cash loans           M            Y   
307495      456239         Cash loans           M            Y   
307496      456240    Revolving loans           F            N   
307497      456241         Cash loans           F            Y   
307498      456242         Cash loans           M            Y   
307499      456243         Cash loans           F            N   
307500      456244         Cash loans           F            N   
307501      456245         Cash loans           F            N   
307502      456246         Cash loans           F            N   
307503      456247         Cash loans           F            N   
307504      456248         Cash loans           F            N   
307505      456249         Cash loans           F            N   
307506      456251         Cash loans           M            N   
307507      456252         Cash loans           F            N   
307508      456253         Cash loans           F            N   
307509      456254         Cash loans           

0         1
1         0
2         0
3         0
4         0
5         0
6         0
7         0
8         0
9         0
10        0
11        0
12        0
13        0
14        0
15        0
16        0
17        0
18        0
19        0
20        0
21        0
22        0
23        0
24        0
25        0
26        1
27        0
28        0
29        0
         ..
307481    1
307482    0
307483    0
307484    0
307485    0
307486    0
307487    0
307488    0
307489    1
307490    0
307491    0
307492    0
307493    0
307494    0
307495    0
307496    0
307497    0
307498    0
307499    0
307500    0
307501    0
307502    0
307503    0
307504    0
307505    0
307506    0
307507    0
307508    0
307509    1
307510    0
Name: TARGET, Length: 307511, dtype: int64

In [5]:
def missing_data(df):
    total = pd.DataFrame(df.isnull().sum().sort_values(ascending = False), columns=['Total'])
    total = total[total['Total'] > 0]
    total['Percent'] = total['Total']/df.shape[0] * 100
    return total

missing_data(df)

Total    Percent
COMMONAREA_MEDI               214865  69.872297
COMMONAREA_AVG                214865  69.872297
COMMONAREA_MODE               214865  69.872297
NONLIVINGAPARTMENTS_MODE      213514  69.432963
NONLIVINGAPARTMENTS_MEDI      213514  69.432963
NONLIVINGAPARTMENTS_AVG       213514  69.432963
FONDKAPREMONT_MODE            210295  68.386172
LIVINGAPARTMENTS_MEDI         210199  68.354953
LIVINGAPARTMENTS_MODE         210199  68.354953
LIVINGAPARTMENTS_AVG          210199  68.354953
FLOORSMIN_MEDI                208642  67.848630
FLOORSMIN_MODE                208642  67.848630
FLOORSMIN_AVG                 208642  67.848630
YEARS_BUILD_MEDI              204488  66.497784
YEARS_BUILD_AVG               204488  66.497784
YEARS_BUILD_MODE              204488  66.497784
OWN_CAR_AGE                   202929  65.990810
LANDAREA_MODE                 182590  59.376738
LANDAREA_AVG                  182590  59.376738
LANDAREA_MEDI                 182590  59.376738
BASEMENTAREA_MEDI             179943  58.515956
BASEMENTAREA_AVG              179943  58.515956
BASEMENTAREA_MODE             179943  58.515956
EXT_SOURCE_1                  173378  56.381073
NONLIVINGAREA_MEDI            169682  55.179164
NONLIVINGAREA_AVG             169682  55.179164
NONLIVINGAREA_MODE            169682  55.179164
ELEVATORS_MODE                163891  53.295980
ELEVATORS_AVG                 163891  53.295980
ELEVATORS_MEDI                163891  53.295980
...                              ...        ...
LIVINGAREA_MEDI               154350  50.193326
LIVINGAREA_MODE               154350  50.193326
LIVINGAREA_AVG                154350  50.193326
HOUSETYPE_MODE                154297  50.176091
FLOORSMAX_MODE                153020  49.760822
FLOORSMAX_MEDI                153020  49.760822
FLOORSMAX_AVG                 153020  49.760822
YEARS_BEGINEXPLUATATION_MEDI  150007  48.781019
YEARS_BEGINEXPLUATATION_AVG   150007  48.781019
YEARS_BEGINEXPLUATATION_MODE  150007  48.781019
TOTALAREA_MODE                148431  48.268517
EMERGENCYSTATE_MODE           145755  47.398304
OCCUPATION_TYPE                96391  31.345545
EXT_SOURCE_3                   60965  19.825307
AMT_REQ_CREDIT_BUREAU_QRT      41519  13.501631
AMT_REQ_CREDIT_BUREAU_YEAR     41519  13.501631
AMT_REQ_CREDIT_BUREAU_WEEK     41519  13.501631
AMT_REQ_CREDIT_BUREAU_MON      41519  13.501631
AMT_REQ_CREDIT_BUREAU_DAY      41519  13.501631
AMT_REQ_CREDIT_BUREAU_HOUR     41519  13.501631
NAME_TYPE_SUITE                 1292   0.420148
OBS_30_CNT_SOCIAL_CIRCLE        1021   0.332021
OBS_60_CNT_SOCIAL_CIRCLE        1021   0.332021
DEF_60_CNT_SOCIAL_CIRCLE        1021   0.332021
DEF_30_CNT_SOCIAL_CIRCLE        1021   0.332021
EXT_SOURCE_2                     660   0.214626
AMT_GOODS_PRICE                  278   0.090403
AMT_ANNUITY                       12   0.003902
CNT_FAM_MEMBERS                    2   0.000650
DAYS_LAST_PHONE_CHANGE             1   0.000325

[67 rows x 2 columns]

In [6]:
# 全変全一括変換用関数
def dummy_variable_all(df, sort_frequence=True, dropna=True, map_show=False):
    """
    オブジェクト型の変数を数値へ一括変換する処理。
    
    Parameters
    ----------------
    df : pd.DataFrame
      データセットのDataFrame。
    sort_frequence : bool
      数値変換の順番を頻度順にするか
    dropna : bool
      nullを同様に数値へ変換するか
    map_show : bool
      変換するname mapを標準出力するか

    Returns
    ----------------
    df : pd.DataFrame
    """
    
    for column_name in df:
        if np.dtype(df[column_name]) == 'object':
            if sort_frequence:
                unique_value = df[column_name].value_counts(dropna=dropna).sort_values(ascending=False).index
            else:
                unique_value = df[column_name].value_counts(dropna=dropna).sort_index().index

            name_map = {}
            for i, value in enumerate(unique_value):
                name_map[value] = i
            if map_show:
                print('column_name :', column_name)
                print('replace :', name_map)
            df[column_name] = df[column_name].map(name_map)
        else:
            pass
    return df

In [7]:
df2 = dummy_variable_all(df, dropna=False, map_show=True)

column_name : NAME_CONTRACT_TYPE
replace : {'Cash loans': 0, 'Revolving loans': 1}
column_name : CODE_GENDER
replace : {'F': 0, 'M': 1, 'XNA': 2}
column_name : FLAG_OWN_CAR
replace : {'N': 0, 'Y': 1}
column_name : FLAG_OWN_REALTY
replace : {'Y': 0, 'N': 1}
column_name : NAME_TYPE_SUITE
replace : {'Unaccompanied': 0, 'Family': 1, 'Spouse, partner': 2, 'Children': 3, 'Other_B': 4, nan: 5, 'Other_A': 6, 'Group of people': 7}
column_name : NAME_INCOME_TYPE
replace : {'Working': 0, 'Commercial associate': 1, 'Pensioner': 2, 'State servant': 3, 'Unemployed': 4, 'Student': 5, 'Businessman': 6, 'Maternity leave': 7}
column_name : NAME_EDUCATION_TYPE
replace : {'Secondary / secondary special': 0, 'Higher education': 1, 'Incomplete higher': 2, 'Lower secondary': 3, 'Academic degree': 4}
column_name : NAME_FAMILY_STATUS
replace : {'Married': 0, 'Single / not married': 1, 'Civil marriage': 2, 'Separated': 3, 'Widow': 4, 'Unknown': 5}
column_name : NAME_HOUSING_TYPE
replace : {'House / apartment': 

In [8]:
df2.fillna(df2.mean(), inplace=True)
df2.head()

SK_ID_CURR  TARGET  NAME_CONTRACT_TYPE  CODE_GENDER  FLAG_OWN_CAR  \
0      100002       1                   0            1             0   
1      100003       0                   0            0             0   
2      100004       0                   1            1             1   
3      100006       0                   0            0             0   
4      100007       0                   0            1             0   

   FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0                0             0          202500.0    406597.5      24700.5   
1                1             0          270000.0   1293502.5      35698.5   
2                0             0           67500.0    135000.0       6750.0   
3                0             0          135000.0    312682.5      29686.5   
4                0             0          121500.0    513000.0      21865.5   

              ...              FLAG_DOCUMENT_18  FLAG_DOCUMENT_19  \
0             ...                             0                 0   
1             ...                             0                 0   
2             ...                             0                 0   
3             ...                             0                 0   
4             ...                             0                 0   

   FLAG_DOCUMENT_20  FLAG_DOCUMENT_21  AMT_REQ_CREDIT_BUREAU_HOUR  \
0                 0                 0                    0.000000   
1                 0                 0                    0.000000   
2                 0                 0                    0.000000   
3                 0                 0                    0.006402   
4                 0                 0                    0.000000   

   AMT_REQ_CREDIT_BUREAU_DAY  AMT_REQ_CREDIT_BUREAU_WEEK  \
0                      0.000                    0.000000   
1                      0.000                    0.000000   
2                      0.000                    0.000000   
3                      0.007                    0.034362   
4                      0.000                    0.000000   

   AMT_REQ_CREDIT_BUREAU_MON  AMT_REQ_CREDIT_BUREAU_QRT  \
0                   0.000000                   0.000000   
1                   0.000000                   0.000000   
2                   0.000000                   0.000000   
3                   0.267395                   0.265474   
4                   0.000000                   0.000000   

   AMT_REQ_CREDIT_BUREAU_YEAR  
0                    1.000000  
1                    0.000000  
2                    0.000000  
3                    1.899974  
4                    0.000000  

[5 rows x 122 columns]

In [9]:
missing_data(df2)

Empty DataFrame
Columns: [Total, Percent]
Index: []

### 【問題1】クロスバリデーション  
事前学習期間では検証データをはじめに分割しておき、それに対して指標値を計算することで検証を行っていました。（ホールドアウト法）しかし、分割の仕方により精度は変化します。実践的には クロスバリデーション（交差検証） を行います。分割を複数回行い、それぞれに対して学習と検証を行う方法です。複数回の分割のためにscikit-learnにはKFoldクラスが用意されています。


事前学習期間の課題で作成したベースラインモデルに対してKFoldクラスによるクロスバリデーションを行うコードを作成し実行してください。


[sklearn.model_selection.KFold — scikit-learn 0.21.3 documentation](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html#sklearn.model_selection.KFold)

In [10]:
X = df2.loc[:, ['AMT_INCOME_TOTAL', 'AMT_CREDIT', 'AMT_ANNUITY', 'AMT_GOODS_PRICE', 
       'REGION_POPULATION_RELATIVE', 'DAYS_BIRTH', 'DAYS_EMPLOYED',
       'DAYS_REGISTRATION', 'DAYS_ID_PUBLISH', 'HOUR_APPR_PROCESS_START',
       'ORGANIZATION_TYPE', 'EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3',
       'DAYS_LAST_PHONE_CHANGE']]
y = df2.loc[:, 'TARGET']

from sklearn.model_selection import KFold
kf = KFold(n_splits=5)
print(kf)
for train_index, test_index in kf.split(X):
    print("TRAIN:", train_index, "TEST:", test_index)
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]

KFold(n_splits=5, random_state=None, shuffle=False)
TRAIN: [ 61503  61504  61505 ... 307508 307509 307510] TEST: [    0     1     2 ... 61500 61501 61502]
TRAIN: [     0      1      2 ... 307508 307509 307510] TEST: [ 61503  61504  61505 ... 123002 123003 123004]
TRAIN: [     0      1      2 ... 307508 307509 307510] TEST: [123005 123006 123007 ... 184504 184505 184506]
TRAIN: [     0      1      2 ... 307508 307509 307510] TEST: [184507 184508 184509 ... 246006 246007 246008]
TRAIN: [     0      1      2 ... 246006 246007 246008] TEST: [246009 246010 246011 ... 307508 307509 307510]


[scikit-learn を用いた交差検証（Cross-validation）とハイパーパラメータのチューニング（grid search）](https://qiita.com/tomov3/items/039d4271ed30490edf7b)

In [11]:
#一般的な（？）クロスバリデーション

# 分割〜学習〜評価を一気にする関数
from sklearn.model_selection import cross_val_score

clf = RandomForestClassifier()

print('Cross-Validation scores: {}'.format(cross_val_score(clf, X, y, scoring="roc_auc")))
print('Average score: {}'.format(np.mean(cross_val_score(clf, X, y, scoring="roc_auc"))))
print('Cross-Validation scores: {}'.format(cross_val_score(clf, X, y)))
print('Average score: {}'.format(np.mean(cross_val_score(clf, X, y))))

Cross-Validation scores: [0.64837206 0.63741712 0.63815092]
Average score: 0.638702965806674
Cross-Validation scores: [0.91752517 0.91795442 0.91795362]
Average score: 0.91796391099377


In [12]:
#KFoldクラスによるクロスバリデーション
from sklearn.model_selection import KFold
from sklearn.model_selection import StratifiedKFold

# 単純なk 割交差検証(一般的に回帰で使用)
kf = KFold(n_splits=3)
print('Cross-validation scores: \n{}'.format(cross_val_score(clf, X, y, cv=kf, scoring="roc_auc")))# 評価指標を指定できる
# 層化k分割交差検証（一般的に分類で使用）
stratifiedkfold = StratifiedKFold(n_splits=3)
print('Cross-validation scores: \n{}'.format(cross_val_score(clf, X, y, cv=stratifiedkfold, scoring="roc_auc")))

Cross-validation scores: 
[0.64317837 0.64017269 0.64248994]
Cross-validation scores: 
[0.64414943 0.6384211  0.64481406]


***

### 【問題2】グリッドサーチ  
これまで分類器のパラメータには触れず、デフォルトの設定を使用していました。パラメータの詳細は今後のSprintで学んでいくことになります。機械学習の前提として、パラメータは状況に応じて最適なものを選ぶ必要があります。最適なパラメータを探していくことを パラメータチューニング と呼びます。パラメータチューニングをある程度自動化する単純な方法としては グリッドサーチ があります。


scikit-learnのGridSearchCVを使い、グリッドサーチを行うコードを作成してください。そして、ベースラインモデルに対して何らかしらのパラメータチューニングを行なってください。どのパラメータをチューニングするかは、使用した手法の公式ドキュメントを参考にしてください。


[sklearn.model_selection.GridSearchCV — scikit-learn 0.21.3 documentation](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html)


GridSearchCVクラスには引数としてモデル、探索範囲、さらにクロスバリデーションを何分割で行うかを与えます。クロスバリデーションの機能も含まれているため、これを使用する場合はKFoldクラスを利用する必要はありません。

[ランダムフォレストで特徴量の重要度を評価する](http://aiweeklynews.com/archives/50653819.html)  

ランダムフォレストもハイパーパラメータの設定が必要です。ただランダムフォレストの場合、デフォルトでも良い精度がわりと出るのであまりこだわらなくても大丈夫です。こだわる人は、いろいろ設定してグリッドサーチしましょう。

- n_estimators　→デフォルト値１０。決定木の数。
- max_features　→デフォルト値は特徴量数の平方根。決定木の特徴量の数。大きいと似たような決定木が増える、小さいと決定木がばらけるがデータに適合できにくくなる。
- random_state　→デフォルト値なし。乱数ジェネレータをどの数値から始めるか。前回と異なるモデルを得たい場合は数値を変更すると良いです。
- max_depth　→デフォルト値なし。決定木の最大の深さ。
- min_samples_split　→デフォルト値2。木を分割する際のサンプル数の最小数。

In [13]:
#交差検証を用いたグリッドサーチ
from sklearn.model_selection import GridSearchCV

param_grid = {'n_estimators': [5, 10, 15],  'max_depth' : [5, 10]}

grid_search = GridSearchCV(clf, param_grid, cv=5) # 5分割考査検証

#パラメータの過剰適合を防ぐために
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

# fit 関数を呼ぶことで交差検証とグリッドサーチがどちらも実行される
grid_search.fit(X_train, y_train)

print('Test set score: {}'.format(grid_search.score(X_test, y_test)))
print('Best parameters: {}'.format(grid_search.best_params_))
print('Best cross-validation: {}'.format(grid_search.best_score_))
kf = KFold(n_splits=3)
print('Cross-validation scores（roc_auc）: \n{}'.format(cross_val_score(grid_search, X, y, cv=kf, scoring="roc_auc")))
print('Cross-validation scores　ave（roc_auc）: \n{}'.format(np.mean(cross_val_score(grid_search, X, y, cv=kf, scoring="roc_auc"))))

Test set score: 0.920835609667265
Best parameters: {'max_depth': 10, 'n_estimators': 15}
Best cross-validation: 0.9188190761946469
Cross-validation scores（roc_auc）: 
[0.72522558 0.70615692 0.72976038]
Cross-validation scores　ave（roc_auc）: 
0.728224939925572


In [14]:
clf2 = RandomForestClassifier(max_depth=10, n_estimators=15)

# 単純なk 割交差検証(一般的に回帰で使用)
kf = KFold(n_splits=3)
print('Cross-validation scores（roc_auc）: \n{}'.format(cross_val_score(clf2, X, y, cv=kf, scoring="roc_auc")))
print('Cross-validation scores　ave（roc_auc）: \n{}'.format(np.mean(cross_val_score(clf2, X, y, cv=kf, scoring="roc_auc"))))

Cross-validation scores（roc_auc）: 
[0.72755285 0.72692755 0.73269149]
Cross-validation scores　ave（roc_auc）: 
0.7303001935716932


### 【問題3】Kaggle Notebooksからの調査
KaggleのNotebooksから様々なアイデアを見つけ出して、列挙してください。


- Light Gradient Boosting Machineを使ってみる

### 【問題4】高い汎化性能のモデル作成
問題3で見つけたアイデアと、独自のアイデアを組み合わせ高い汎化性能のモデル作りを進めてください。


その過程として、何を行うことで、クロスバリデーションの結果がどの程度変化したかを表にまとめてください。

- Light Gradient Boosting Machineを使ってみる

In [15]:
#パラメータ調整無し
import lightgbm as lgb
lgb = lgb.LGBMClassifier()

# 単純なk 割交差検証(一般的に回帰で使用)
kf = KFold(n_splits=3)
print('Cross-validation scores: \n{}'.format(cross_val_score(lgb, X, y, cv=kf, scoring="roc_auc")))
print('Cross-validation ave.scores: \n{}'.format(np.mean(cross_val_score(lgb, X, y, cv=kf, scoring="roc_auc"))))

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/lightgbm/__init__.py:48: UserWarning: Starting from version 2.2.1, the library file in distribution wheels for macOS is built by the Apple Clang (Xcode_8.3.3) compiler.
This means that in case of installing LightGBM from PyPI via the ``pip install lightgbm`` command, you don't need to install the gcc compiler anymore.
Instead of that, you need to install the OpenMP library, which is required for running LightGBM on the system with the Apple Clang compiler.
You can install the OpenMP library by the following command: ``brew install libomp``.
  "You can install the OpenMP library by the following command: ``brew install libomp``.", UserWarning)


Cross-validation scores: 
[0.74612082 0.7428588  0.747986  ]
Cross-validation ave.scores: 
0.7456552111305287


[【機械学習】ハイパーパラメータチューニングの仕方](https://qiita.com/R1ck29/items/50ba7fa5afa49e334a8f)  
[LightGBM パラメータ調整](https://lightgbm.readthedocs.io/en/latest/Parameters-Tuning.html)

In [16]:
#交差検証を用いたグリッドサーチ
from sklearn.model_selection import GridSearchCV

param_grid = {"num_leaves":[7, 20, 30, 40], \
              "min_data_in_leaf":[30, 100, 500, 1000], \
             "learning_rate":[0.05, 0.1, 0.15, 0.2]}

#パラメータの過剰適合を防ぐために
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

grid_search = GridSearchCV(lgb, param_grid, cv=3)

# fit 関数を呼ぶことで交差検証とグリッドサーチがどちらも実行される
grid_search.fit(X_train, y_train)

print('Test set score: {}'.format(grid_search.score(X_test, y_test)))
print('Best parameters: {}'.format(grid_search.best_params_))
print('Best cross-validation: {}'.format(grid_search.best_score_))

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-pac

Test set score: 0.9209396706469991
Best parameters: {'learning_rate': 0.1, 'min_data_in_leaf': 500, 'num_leaves': 40}
Best cross-validation: 0.9191139169156192


/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


In [17]:
#訓練モデル
grid_search.best_estimator_

LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
        importance_type='split', learning_rate=0.1, max_depth=-1,
        min_child_samples=20, min_child_weight=0.001, min_data_in_leaf=500,
        min_split_gain=0.0, n_estimators=100, n_jobs=-1, num_leaves=40,
        objective=None, random_state=None, reg_alpha=0.0, reg_lambda=0.0,
        silent=True, subsample=1.0, subsample_for_bin=200000,
        subsample_freq=0)

In [18]:
#グリッドサーチの結果
results = pd.DataFrame(grid_search.cv_results_)
display(results.head())

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/utils/deprecation.py:122: FutureWarning: You are accessing a training score ('split0_train_score'), which will not be available by default any more in 0.21. If you need training scores, please set return_train_score=True
  warnings.warn(*warn_args, **warn_kwargs)
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/utils/deprecation.py:122: FutureWarning: You are accessing a training score ('split1_train_score'), which will not be available by default any more in 0.21. If you need training scores, please set return_train_score=True
  warnings.warn(*warn_args, **warn_kwargs)
/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/utils/deprecation.py:122: FutureWarning: You are accessing a training score ('split2_train_score'), which will not be available by default any more in 0.21. If you need training scores, please set return_train_score=True

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.697443      0.150483         0.089242        0.019630   
1       0.677775      0.093969         0.090973        0.001963   
2       0.761945      0.106067         0.103830        0.009161   
3       0.675166      0.062264         0.104675        0.002553   
4       0.418167      0.015610         0.084557        0.010303   

  param_learning_rate param_min_data_in_leaf param_num_leaves  \
0                0.05                     30                7   
1                0.05                     30               20   
2                0.05                     30               30   
3                0.05                     30               40   
4                0.05                    100                7   

                                              params  split0_test_score  \
0  {'learning_rate': 0.05, 'min_data_in_leaf': 30...           0.918858   
1  {'learning_rate': 0.05, 'min_data_in_leaf': 30...           0.919041   
2  {'learning_rate': 0.05, 'min_data_in_leaf': 30...           0.918976   
3  {'learning_rate': 0.05, 'min_data_in_leaf': 30...           0.918923   
4  {'learning_rate': 0.05, 'min_data_in_leaf': 10...           0.918858   

   split1_test_score  split2_test_score  mean_test_score  std_test_score  \
0           0.918806           0.918870         0.918845        0.000028   
1           0.918923           0.919013         0.918993        0.000050   
2           0.918871           0.918909         0.918919        0.000043   
3           0.918871           0.918883         0.918893        0.000022   
4           0.918806           0.918870         0.918845        0.000028   

   rank_test_score  split0_train_score  split1_train_score  \
0               53            0.918838            0.918871   
1               22            0.919183            0.919365   
2               40            0.919346            0.919573   
3               47            0.919534            0.919807   
4               53            0.918845            0.918871   

   split2_train_score  mean_train_score  std_train_score  
0            0.918858          0.918856         0.000013  
1            0.919158          0.919235         0.000092  
2            0.919385          0.919435         0.000099  
3            0.919502          0.919615         0.000137  
4            0.918871          0.918862         0.000012

In [19]:
#パラメータ調整あり
import lightgbm as lgb
lgb2 = lgb.LGBMClassifier(min_data_in_leaf=500, num_leaves=40)

# 単純なk 割交差検証(一般的に回帰で使用)
kf = KFold(n_splits=3)
print('Cross-validation scores: \n{}'.format(cross_val_score(lgb2, X, y, cv=kf, scoring="roc_auc")))
print('Cross-validation ave.scores: \n{}'.format(np.mean(cross_val_score(lgb2, X, y, cv=kf, scoring="roc_auc"))))

Cross-validation scores: 
[0.74779077 0.74581573 0.75124526]
Cross-validation ave.scores: 
0.7482839181602641


|---| ROC_AUCスコア（平均） |
|-|-|
|ランダムフォレスト（パラメータ調整無し）                                               |0.722755418477607|
|ランダムフォレスト（max_depth=10, n_estimators=15）  　                        |0.7295396294662019|
|LightGBM（パラメータ調整無し）                                                       |0.7456552111305287|
|LightGBM（min_data_in_leaf=500, num_leaves=40）                           |0.7482839181602641|

### 【問題5】最終的なモデルの選定
最終的にこれは良いというモデルを選び、推定した結果をKaggleに提出してスコアを確認してください。どういったアイデアを取り入れ、どの程度のスコアになったかを記載してください。

In [21]:
df_test = pd.read_csv('application_test.csv')
display(df_test)

SK_ID_CURR NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
0          100001         Cash loans           F            N               Y   
1          100005         Cash loans           M            N               Y   
2          100013         Cash loans           M            Y               Y   
3          100028         Cash loans           F            N               Y   
4          100038         Cash loans           M            Y               N   
5          100042         Cash loans           F            Y               Y   
6          100057         Cash loans           M            Y               Y   
7          100065         Cash loans           M            N               Y   
8          100066         Cash loans           F            N               Y   
9          100067         Cash loans           F            Y               Y   
10         100074         Cash loans           F            N               Y   
11         100090         Cash loans           F            N               Y   
12         100091         Cash loans           F            N               Y   
13         100092         Cash loans           F            Y               Y   
14         100106    Revolving loans           M            N               Y   
15         100107         Cash loans           M            Y               Y   
16         100109         Cash loans           F            Y               Y   
17         100117         Cash loans           M            Y               Y   
18         100128         Cash loans           F            Y               Y   
19         100141         Cash loans           F            Y               Y   
20         100150         Cash loans           F            N               Y   
21         100168         Cash loans           F            N               Y   
22         100169         Cash loans           F            N               Y   
23         100170         Cash loans           M            N               Y   
24         100171         Cash loans           M            Y               Y   
25         100172         Cash loans           F            N               Y   
26         100184         Cash loans           F            N               Y   
27         100187         Cash loans           F            N               Y   
28         100212         Cash loans           F            Y               N   
29         100222         Cash loans           F            Y               Y   
...           ...                ...         ...          ...             ...   
48714      455963         Cash loans           F            N               N   
48715      455965         Cash loans           M            Y               Y   
48716      456007         Cash loans           M            N               Y   
48717      456008         Cash loans           F            N               Y   
48718      456009         Cash loans           F            N               Y   
48719      456010         Cash loans           F            N               Y   
48720      456011         Cash loans           F            Y               Y   
48721      456013         Cash loans           F            N               Y   
48722      456028         Cash loans           M            N               Y   
48723      456058         Cash loans           F            N               N   
48724      456111         Cash loans           F            N               Y   
48725      456114         Cash loans           F            N               N   
48726      456115         Cash loans           M            N               Y   
48727      456116         Cash loans           F            N               N   
48728      456119         Cash loans           F            N               Y   
48729      456120         Cash loans           F            N               N   
48730      456122         Cash loans           M            Y               N   
48731      456123         Cash loans  

In [22]:
df_test2 = dummy_variable_all(df_test, dropna=False, map_show=True)

column_name : NAME_CONTRACT_TYPE
replace : {'Cash loans': 0, 'Revolving loans': 1}
column_name : CODE_GENDER
replace : {'F': 0, 'M': 1}
column_name : FLAG_OWN_CAR
replace : {'N': 0, 'Y': 1}
column_name : FLAG_OWN_REALTY
replace : {'Y': 0, 'N': 1}
column_name : NAME_TYPE_SUITE
replace : {'Unaccompanied': 0, 'Family': 1, 'Spouse, partner': 2, nan: 3, 'Children': 4, 'Other_B': 5, 'Other_A': 6, 'Group of people': 7}
column_name : NAME_INCOME_TYPE
replace : {'Working': 0, 'Commercial associate': 1, 'Pensioner': 2, 'State servant': 3, 'Student': 4, 'Businessman': 5, 'Unemployed': 6}
column_name : NAME_EDUCATION_TYPE
replace : {'Secondary / secondary special': 0, 'Higher education': 1, 'Incomplete higher': 2, 'Lower secondary': 3, 'Academic degree': 4}
column_name : NAME_FAMILY_STATUS
replace : {'Married': 0, 'Single / not married': 1, 'Civil marriage': 2, 'Separated': 3, 'Widow': 4}
column_name : NAME_HOUSING_TYPE
replace : {'House / apartment': 0, 'With parents': 1, 'Municipal apartment': 2

In [23]:
df_test2.fillna(df_test2.mean(), inplace=True)
df_test2.head()

SK_ID_CURR  NAME_CONTRACT_TYPE  CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  \
0      100001                   0            0             0                0   
1      100005                   0            1             0                0   
2      100013                   0            1             1                0   
3      100028                   0            0             0                0   
4      100038                   0            1             1                1   

   CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  AMT_GOODS_PRICE  \
0             0          135000.0    568800.0      20560.5         450000.0   
1             0           99000.0    222768.0      17370.0         180000.0   
2             0          202500.0    663264.0      69777.0         630000.0   
3             2          315000.0   1575000.0      49018.5        1575000.0   
4             1          180000.0    625500.0      32067.0         625500.0   

              ...              FLAG_DOCUMENT_18  FLAG_DOCUMENT_19  \
0             ...                             0                 0   
1             ...                             0                 0   
2             ...                             0                 0   
3             ...                             0                 0   
4             ...                             0                 0   

   FLAG_DOCUMENT_20  FLAG_DOCUMENT_21  AMT_REQ_CREDIT_BUREAU_HOUR  \
0                 0                 0                    0.000000   
1                 0                 0                    0.000000   
2                 0                 0                    0.000000   
3                 0                 0                    0.000000   
4                 0                 0                    0.002108   

   AMT_REQ_CREDIT_BUREAU_DAY  AMT_REQ_CREDIT_BUREAU_WEEK  \
0                   0.000000                    0.000000   
1                   0.000000                    0.000000   
2                   0.000000                    0.000000   
3                   0.000000                    0.000000   
4                   0.001803                    0.002787   

   AMT_REQ_CREDIT_BUREAU_MON  AMT_REQ_CREDIT_BUREAU_QRT  \
0                   0.000000                   0.000000   
1                   0.000000                   0.000000   
2                   0.000000                   1.000000   
3                   0.000000                   0.000000   
4                   0.009299                   0.546902   

   AMT_REQ_CREDIT_BUREAU_YEAR  
0                    0.000000  
1                    3.000000  
2                    4.000000  
3                    3.000000  
4                    1.983769  

[5 rows x 121 columns]

In [24]:
X2 = df_test2.loc[:, ['AMT_INCOME_TOTAL', 'AMT_CREDIT', 'AMT_ANNUITY', 'AMT_GOODS_PRICE', 
       'REGION_POPULATION_RELATIVE', 'DAYS_BIRTH', 'DAYS_EMPLOYED',
       'DAYS_REGISTRATION', 'DAYS_ID_PUBLISH', 'HOUR_APPR_PROCESS_START',
       'ORGANIZATION_TYPE', 'EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3',
       'DAYS_LAST_PHONE_CHANGE']]
display(X2)

AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  AMT_GOODS_PRICE  \
0              135000.0    568800.0      20560.5         450000.0   
1               99000.0    222768.0      17370.0         180000.0   
2              202500.0    663264.0      69777.0         630000.0   
3              315000.0   1575000.0      49018.5        1575000.0   
4              180000.0    625500.0      32067.0         625500.0   
5              270000.0    959688.0      34600.5         810000.0   
6              180000.0    499221.0      22117.5         373500.0   
7              166500.0    180000.0      14220.0         180000.0   
8              315000.0    364896.0      28957.5         315000.0   
9              162000.0     45000.0       5337.0          45000.0   
10              67500.0    675000.0      25447.5         675000.0   
11             135000.0    261621.0      16848.0         216000.0   
12             247500.0    296280.0      23539.5         225000.0   
13              90000.0    360000.0      18535.5         360000.0   
14             180000.0    157500.0       7875.0         157500.0   
15             180000.0    296280.0      21690.0         225000.0   
16             202500.0    407520.0      26041.5         360000.0   
17              90000.0    499221.0      22117.5         373500.0   
18             225000.0    431280.0      23526.0         360000.0   
19             175500.0    478498.5      46741.5         454500.0   
20              99000.0    225000.0      19242.0         225000.0   
21             157500.0    266652.0      16443.0         202500.0   
22             135000.0    540000.0      27702.0         540000.0   
23             337500.0   1313212.5      42493.5        1012500.0   
24             157500.0    539100.0      29245.5         450000.0   
25              76500.0    225000.0      12334.5         225000.0   
26             112500.0    256032.0      10759.5         180000.0   
27             225000.0    501363.0      25726.5         418500.0   
28              90000.0    450000.0      23107.5         450000.0   
29             360000.0    945000.0      34074.0         945000.0   
...                 ...         ...          ...              ...   
48714          360000.0    547272.0      43366.5         495000.0   
48715          315000.0    900000.0      57519.0         900000.0   
48716          135000.0    318645.0      16398.0         216000.0   
48717           81000.0    180000.0      13455.0         180000.0   
48718          225000.0    360000.0      19530.0         360000.0   
48719          112500.0    690313.5      35374.5         549000.0   
48720          171000.0    323460.0      23134.5         270000.0   
48721          130500.0    468333.0      37129.5         382500.0   
48722          220500.0    166500.0      13284.0         166500.0   
48723           67500.0    315000.0      16213.5         315000.0   
48724          112500.0    514710.0      17707.5         387000.0   
48725           45000.0    450000.0      24543.0         450000.0   
48726          225000.0   1002339.0      37917.0         846000.0   
48727          180000.0    149256.0      15291.0         135000.0   
48728          247500.0    685386.0      38398.5         585000.0   
48729           81000.0    135000.0      16020.0         135000.0   
48730          157500.0    445095.0      40824.0         405000.0   
48731          112500.0    450000.0      28890.0         450000.0   
48732          189000.0    479700.0      54382.5         450000.0   
48733          135000.0   1047771.0      37764.0         904500.0   
48734          135000.0    835686.0      46660.5         765000.0   
48735           85500.0    109008.0       7411.5          90000.0   
48736          157500.0    500490.0      48888.0         450000.0   
48737          270000.0    360000.0      28570.5         360000.0   
48738          135000.0    252022.5      23112.0         217561.5   
48739          121500.0    412560.0      17473.5         270000.0   
48740

In [25]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
lgb2.fit(X_train, y_train)

y_pred = lgb2.predict_proba(X2)
y_pred

array([[0.96155706, 0.03844294],
       [0.9404525 , 0.0595475 ],
       [0.95696708, 0.04303292],
       ...,
       [0.95459525, 0.04540475],
       [0.92571884, 0.07428116],
       [0.84769625, 0.15230375]])

In [26]:
lgb2.predict(X2)

/Users/itonaoki/.pyenv/versions/anaconda3-5.3.1/lib/python3.7/site-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


array([0, 0, 0, ..., 0, 0, 0])

In [27]:
submit = pd.concat([df_test2['SK_ID_CURR'], pd.Series(y_pred[:, 1])], axis=1)
submit = submit.rename(columns={0: 'TARGET'})
display(submit)

SK_ID_CURR    TARGET
0          100001  0.038443
1          100005  0.059547
2          100013  0.043033
3          100028  0.041162
4          100038  0.130650
5          100042  0.053374
6          100057  0.022728
7          100065  0.049806
8          100066  0.013485
9          100067  0.088634
10         100074  0.073310
11         100090  0.038512
12         100091  0.169716
13         100092  0.127598
14         100106  0.044836
15         100107  0.110624
16         100109  0.044380
17         100117  0.023554
18         100128  0.085135
19         100141  0.019609
20         100150  0.020816
21         100168  0.014831
22         100169  0.029331
23         100170  0.047568
24         100171  0.040972
25         100172  0.050075
26         100184  0.069597
27         100187  0.110476
28         100212  0.055620
29         100222  0.067896
...           ...       ...
48714      455963  0.043955
48715      455965  0.017319
48716      456007  0.336273
48717      456008  0.033422
48718      456009  0.037932
48719      456010  0.062912
48720      456011  0.038741
48721      456013  0.119227
48722      456028  0.133270
48723      456058  0.154981
48724      456111  0.154497
48725      456114  0.056387
48726      456115  0.067977
48727      456116  0.026425
48728      456119  0.033616
48729      456120  0.050460
48730      456122  0.078990
48731      456123  0.046497
48732      456166  0.147407
48733      456167  0.075919
48734      456168  0.050449
48735      456169  0.065549
48736      456170  0.026374
48737      456189  0.139930
48738      456202  0.188929
48739      456221  0.040650
48740      456222  0.052960
48741      456223  0.045405
48742      456224  0.074281
48743      456250  0.152304

[48744 rows x 2 columns]

In [28]:
submit.to_csv('kaggle2.csv', index=False)

モデルはLightGBMを使い、パラメータチューニングを行い、スコアは「0.72811」